modules/ocf/files/ssl/ocf-lets-encrypt

```diff
@@ -0,0 +1,254 @@
+def make_csr(domains, key_path):
+    req = crypto.X509Req()
+    req.get_subject().CN = domains[0]
+
+    with open(key_path, 'rt') as f:
+        key = crypto.load_privatekey(crypto.FILETYPE_PEM, f.read())
+        req.set_pubkey(key)
+
+    req.add_extensions((crypto.X509Extension(
+        b'subjectAltName',
+        False,
+        bytes(', '.join('DNS:' + domain for domain in domains), encoding='ascii'),
+    ),))
+
+    req.sign(key, 'sha256')
+
+    with tempfile.NamedTemporaryFile() as csr:
+        csr.write(crypto.dump_certificate_request(crypto.FILETYPE_PEM, req))
+        csr.flush()
+        yield csr.name
+
+
+def acme_tiny(domains, key_path):
+    with make_csr(domains, key_path) as csr_path:
+        # TODO: use subprocess.run when we get python3.5
+        proc = subprocess.Popen(
+            (
+                'acme-tiny',
+                '--account-key', str(LE_ACCOUNT_KEY),
+                '--csr', csr_path,
+                '--acme-dir', str(ACME_BASE_DIR / '.well-known' / 'acme-challenge'),
+                # '--ca', 'https://acme-staging.api.letsencrypt.org',
+            ),
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        stdout, stderr = proc.communicate()
+    return stdout, stderr, proc.returncode
+
+
+def write_cert(cert, cert_path):
+    """Write certificate file."""
+    with tempfile.NamedTemporaryFile(
+        dir=str(cert_path.parent),
+        prefix='.tmp',
+        delete=False,
+    ) as temp_cert:
+        try:
+            temp_cert.write(cert)
+            # atomic overwrite in case of errors
+            os.chmod(temp_cert.name, 0o644)
+            os.rename(temp_cert.name, str(cert_path))
+        finally:
+            try:
+                os.remove(temp_cert.name)
+            except FileNotFoundError:
+                pass
+
+
+def make_request(domains, cert_path, key, dry_run=False):
+    """Makes a Let's Encrypt certificate request."""
+    for domain in domains:
+        assert re.match(r'^[a-z\-_\.0-9]+$', domain)
+    if not dry_run:
+        try:
+            cert, stderr, returncode = acme_tiny(domains, key)
+            assert returncode == 0, returncode
+            write_cert(cert, cert_path)
+        except:
+            print('An error occured!', file=sys.stderr)
+            print('stderr from acme-tiny:', file=sys.stderr)
+            for line in stderr.decode().split('\n'):
+                print('>', line, file=sys.stderr)
+            raise
+
+
+@contextmanager
+def maybe_start_webserver(dry_run):
+    if not dry_run and socket.socket().connect_ex(('127.0.0.1', 80)) != 0:
+        proc = subprocess.Popen(
+            ('python3', '-m', 'http.server', '80'),
+            cwd=str(ACME_BASE_DIR),
+        )
+
+        # Wait for server to start up before proceeding
+        while socket.socket().connect_ex(('127.0.0.1', 80)) != 0:
+            assert proc.poll() is None, proc.returncode
+
+        yield proc
+
+        proc.terminate()
```

Comment:
Well especially if we change it to `proc.terminate()`, then we may want to wait for the process to actually finish...

Reply:
You mean like call `terminate()`, call `communicate()` with a timeout, and then call `kill()` if it still hasn't finished? If so, then, sure, it doesn't hurt, but I don't see the benefit.